File: fetch_border.dart
import 'dart:convert';
import 'dart:io';

void main() async {
  final url = Uri.parse('https://raw.githubusercontent.com/glynnbird/countriesgeojson/master/vietnam.geojson');
  print('Fetching $url...');
  final httpClient = HttpClient();
  try {
    final request = await httpClient.getUrl(url);
    final response = await request.close();
    final stringData = await response.transform(utf8.decoder).join();
    
    final geojson = json.decode(stringData);
    
    // Some GeoJSON wrap in FeatureCollection
    Map<String, dynamic> geometry;
    if (geojson['type'] == 'FeatureCollection') {
      geometry = geojson['features'][0]['geometry'];
    } else if (geojson['type'] == 'Feature') {
      geometry = geojson['geometry'];
    } else {
      geometry = geojson;
    }
    
    List<dynamic> coords = geometry['coordinates'];
    
    StringBuffer dartCode = StringBuffer();
    dartCode.writeln("import 'package:latlong2/latlong.dart';");
    dartCode.writeln("");
    dartCode.writeln("final List<List<LatLng>> vietnamBorderPolygons = [");
    
    if (geometry['type'] == 'Polygon') {
      coords = [coords];
    }
    
    int polyCount = 0;
    for (var polygon in coords) {
      var ring = polygon[0]; // exterior ring
      dartCode.writeln("  [");
      for (var pt in ring) {
        dartCode.writeln("    LatLng(${pt[1]}, ${pt[0]}),");
      }
      dartCode.writeln("  ],");
      polyCount++;
    }
    dartCode.writeln("];");
    
    File('lib/vietnam_border.dart').writeAsStringSync(dartCode.toString());
    print("Successfully wrote lib/vietnam_border.dart with \$polyCount polygons.");
    
  } catch(e) {
    print('Error: $e');
  } finally {
    httpClient.close();
  }
}

File: fetch_border.js
const fs = require('fs');
const https = require('https');

const url = 'https://raw.githubusercontent.com/glynnbird/countriesgeojson/master/vietnam.geojson';

https.get(url, (res) => {
    let data = '';
    res.on('data', chunk => { data += chunk; });
    res.on('end', () => {
        try {
            const geojson = JSON.parse(data);
            let coords = [];
            
            // Handle Polygon or MultiPolygon
            let type = geojson.geometry ? geojson.geometry.type : geojson.features[0].geometry.type;
            let geometryCoords = geojson.geometry ? geojson.geometry.coordinates : geojson.features[0].geometry.coordinates;

            let dartCode = "import 'package:latlong2/latlong.dart';\n\n";
            dartCode += "final List<List<LatLng>> vietnamBorderPolygons = [\n";

            if (type === 'Polygon') {
                geometryCoords = [geometryCoords];
            }

            // MultiPolygon is [[[ [lng, lat], ... ]]]
            for (let polygon of geometryCoords) {
                // Usually polygon[0] is the exterior ring
                let ring = polygon[0];
                dartCode += "  [\n";
                for (let pt of ring) {
                    dartCode += `    LatLng(${pt[1]}, ${pt[0]}),\n`;
                }
                dartCode += "  ],\n";
            }
            dartCode += "];\n";

            fs.writeFileSync('lib/vietnam_border.dart', dartCode);
            console.log("Successfully wrote lib/vietnam_border.dart with " + geometryCoords.length + " polygons.");
        } catch(e) {
            console.error("Error parsing GeoJSON:", e);
        }
    });
}).on('error', (e) => {
    console.error(e);
});
